### PR TITLE
[SPARK-30679][SQL] REPLACE TABLE can omit the USING clause

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This PR updates the parser rule to allow `REPLACE TABLE` statement to omit the `USING` clause.

This is extracted from https://github.com/apache/spark/pull/27345

### Why are the changes needed?

`CREATE TABLE` can omit the `USING` clause, and `REPLACE TABLE` should be consistent.

### Does this PR introduce any user-facing change?

yes, now it's easier to write `REPLACE TABLE` command.

### How was this patch tested?

a new test

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @rdblue @brkyvz @gengliangwang

**Comment by brkyvz** [comment]
I actually don't like the fact that you can be accidentally changing the format your table is stored in, but if people think that's fine, I won't oppose this change

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 1]
```diff
@@ -2813,7 +2813,9 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
     val (partitioning, bucketSpec, properties, options, location, comment) =
       visitCreateTableClauses(ctx.createTableClauses())
     val schema = Option(ctx.colTypeList()).map(createSchema)
-    val provider = ctx.tableProvider.multipartIdentifier.getText
+    val defaultProvider = conf.defaultDataSourceName
+    val provider =
+      Option(ctx.tableProvider).map(_.multipartIdentifier.getText).getOrElse(defaultProvider)
```
Comment (rdblue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: The default provider should not be set here. The purpose of the statement logical plans is to carry only what was expressed in SQL. Replacing with a default loses the fact that this query had no provider specified.

Instead, this needs to update the statement plan and pass provider as an option.

**Comment by rdblue** [comment]
I'm -1 on this because it incorrectly returns a statement that fills in information not in the user's SQL query. Defaults should be set later.

**Comment by cloud-fan** [comment]
Hi @rdblue , I just remind that, by design the default provider should be determined by the catalog implementation, not Spark.

I think you are right that the statement should hold an option and we shouldn't set the "provider" property when calling `TableCatalog.createTable`. I'll fix REPLACE TABLE here, and we should fix CREATE TABLE as well later.

**Comment by gengliangwang** [comment]
> I actually don't like the fact that you can be accidentally changing the format your table is stored in, but if people think that's fine, I won't oppose this change

I am +1 with @brkyvz

**Comment by gengliangwang** [comment]
retest this please

**Comment by cloud-fan** [comment]
Seems no one is against allowing to omit the USING clause, but the question is what the default value should be. I see two options:
1. let catalog implementation to decide the default value. If the catalog doesn't implement `StagingTableCatalog`, then Spark will simply drop and re-create the table. It's hard for the catalog to retain the provider when `createTable` is called, as the table is already dropped.
2. Spark sets the default value as the provider of the existing table. This makes sure that when the USING clause is omitted, we won't accidentally change the table format.

I'll go with option 2.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala [hunk 2]
```diff
@@ -49,26 +49,6 @@ class DDLParserSuite extends AnalysisTest {
     comparePlans(parsePlan(sql), expected, checkAnalysis = false)
   }
 
-  test("SPARK-30098: create table without provider should " +
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala]: The default provider shouldn't be set by the parser. The newly added test case will test the default provider.

**Comment by cloud-fan** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by cloud-fan** [comment]
It's not a breaking change to allow to omit the provider in REPLACE TABLE, so I'll leave it to 3.1.

The CREATE TABLE behavior should be fixed before 3.0: https://github.com/apache/spark/pull/27650

**Pull request closed without merging**